allen-cell-animated/timelapse-colorizer
# feature: Draw motion vectors

Problem
=======
Closes #42, "visualize XY movement delta." This PR gets a functional implementation of vector lines into TFE! They can now be toggled on and off, and parameters like smoothing and scaling can be edited from a new settings area.

There are a number of spin-off issues that I've created that will be follow-ups to this PR.
- Sync vector state to the URL: #471
- Handle vector scaling using a vertex shader: #470
- Remove collapse in SettingsTab: #472

Solution
========
This is a big one, but I think it's split into relatively understandable chunks! I'll be sending a meeting invite to whoever's assigned to this PR so I can walk through it.

- Added math logic for calculating the motion delta of all objects, with smoothing and thresholding applied.
- Added a new async worker task for calculation of motion deltas.
- Added a React hook for managing/syncing motion delta state.
- Added a settings area for configuring vector visualization.
- Added the new `VectorField` class, which manages a Three.JS line used for rendering the vector arrows.
  - `VectorField` calculates the vertices for ALL objects across ALL timepoints into a big buffer, then only renders slices of it based on the current frame.

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Open the PR preview link here: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-469
2. Go to the settings tab and turn on vector arrows
3. Try changing frames, playing through time, and adjusting vector settings.
4. The vector arrows should stay constant size with zooming also.

Screenshots (optional):
-----------------------

https://github.com/user-attachments/assets/b89c6c8e-e066-4e0e-b7d7-4ff14e369f72



Keyfiles (delete if not relevant):
-----------------------
SUGGESTED REVIEW ORDER:
1. `math_utils.ts` -> Deltas are calculated here!
5. `types.ts` -> Vector configuration
6. `SharedWorkerPool.ts` + `worker.ts` -> Delta calculations can be triggered asynchronously on workers.
7. `Viewer` + `react_utils.ts` -> Manages synchronizing react state and delta calculation
8. `ColorizeCanvas` + `VectorField` -> actually renders the vectors to the screen
9. `VectoFieldSettings` -> new settings area
